### PR TITLE
Pagination with router params

### LINKED ISSUE #3868: Pagination component does not keep state when navigating away from it
**Describe the bug**

When the user navigates away from a listing that has pagination, the listing block doesn't keep the state on which was rendered, and if the user goes back to the listing block, the pagination state is recalculated and rendered again from the 1st page.

**To Reproduce**

1. Add a Listing or a Search block and configure it in such a way that it has more than 1 page in the results.
2. Browse the page with the block you have just created and go to the 2nd page
3. Click in a result item and go to its page
4. Use the back button in the browser
5. The listing block results are rendered again but they show the 1st page

**Expected behavior**
After going back I expect to be in the same page that I was before I left the page

**Additional context**
There is no querystring like ?page=1, or b_start=20 or something like that that could help on getting the actual page the user was on.

### PR DESCRIPTION
**Pull request by erral**
Fixes #3868 and #4692

### CODE REVIEW

**Comment by erral** [comment]
@sneridagh we have reworked this PR and added cypress tests to show that the `?page=XX` functionality is working. We test clicking on the pagination numbers, going back, reloading the page and going to some other page too.

In the meantime we have fixed the `withQuerystringResults` function, which wasn't checking for the request being loaded, but being "not loading" which created additional queries to the endpoint.

**Comment by ionlizarazu** [comment]
This is the cypress tests resulting video for the ListingBlock

https://user-images.githubusercontent.com/5443301/233557565-51fce414-f1fa-48d5-bcea-625ac4837f93.mp4

**Comment by sneridagh** [comment]
@erral @ionlizarazu I've found a problem, see video. When I create a page, then go back to the listing, it's still empty :( so the query is not being triggered on route change for some reason.

https://user-images.githubusercontent.com/486927/236615730-ec90693c-b961-47e8-adc4-d12c83911ff5.mov

**Comment by ionlizarazu** [comment]
> @erral @ionlizarazu I've found a problem, see video. When I create a page, then go back to the listing, it's still empty :( so the query is not being triggered on route change for some reason. I've tried on demo and works well.
> 
>  Screen.Recording.2023-05-06.at.11.23.48.mov

I've added a cypress test for this and fixed it.

**Comment by robgietema** [comment]
How does this work when I add multiple listing blocks on 1 page?

**Comment by ionlizarazu** [comment]
> How does this work when I add multiple listing blocks on 1 page?

https://cloud.cypress.io/projects/hvviu4/runs/df7a2ee0-472a-456a-8191-814a1b6c3b35/test-results/5fa7949f-3218-4883-a052-3adf784c06d4/video

**Review by robgietema — APPROVED** [review]

**Review by sneridagh — APPROVED** [review]

File: src/helpers/Utils/usePagination.js [hunk 1]
```diff
@@ -1,25 +1,83 @@
...
+import { useSelector } from 'react-redux';
+import { slugify } from '@plone/volto/helpers/Utils/Utils';
+
+/**
+ * @function useCreatePageQueryStringKey
+ * @description A hook that creates a key with an id if there are multiple blocks with pagination.
+ * @returns {string} Example: page || page_012345678
+ */
+const useCreatePageQueryStringKey = (id) => {
+  const blockTypesWithPagination = ['search', 'listing'];
```
Comment (davisagli) [inline src/helpers/Utils/usePagination.js:13]: Maybe this should come from the config so that addon blocks can use pagination.

**Comment by davisagli** [comment]
I tried editing a search block and changing the number of items per page, then saving. The items shown per page updated correctly in edit mode, but after I saved, the view mode was still showing the old number of items per page until I reloaded.

File: src/helpers/Utils/usePagination.js [hunk 2]
```diff
@@ -1,25 +1,83 @@
...
+      };
+      delete newParams[pageQueryStringKey];
+      history.replace({ search: qs.stringify(newParams) });
+      queryRef.current = qs.parse(location.search)?.query;
+    } else {
+      setCurrentPageState(
+        parseInt(
+          qs.parse(location.search)?.[pageQueryStringKey] || defaultPage,
+        ),
+      );
```
Comment (davisagli) [inline src/helpers/Utils/usePagination.js:74]: The parsing of the page from the querystring needs to handle a couple edge cases better:
1. I navigated to `?page=x` and the pagination showed `NaN`
2. I navigated to `?page=-4` and the block showed no results.

If the value isn't a positive integer, this should probably just use the default page.

**Comment by ionlizarazu** [comment]
> I tried editing a search block and changing the number of items per page, then saving. The items shown per page updated correctly in edit mode, but after I saved, the view mode was still showing the old number of items per page until I reloaded.

This is not an issue we are handling in this PR and it happens in the actual Volto version. Can we manage it in another issue or PR?

File: src/helpers/Utils/usePagination.js (hunk 1 above)
Comment (ionlizarazu) [reply src/helpers/Utils/usePagination.js:13]: You mean the 'blockTypesWithPagination'?

Comment (davisagli) [reply src/helpers/Utils/usePagination.js:13]: @ionlizarazu Yes. But it can wait for a future PR, rather than delaying this one.

**Comment by sneridagh** [comment]
@ionlizarazu I also think it's ready, I'll try to merge it this week.

**Pull request merged**